mmcv/cnn/bricks/transformer.py

```diff
@@ -0,0 +1,433 @@
+    """
+
+    def __init__(
+            self,
+            attn_cfgs=None,
+            feedforward_channels=None,
+            ffn_dropout=0.0,
+            operation_order=None,
+            act_cfg=dict(type='ReLU', inplace=True),
+            norm_cfg=dict(type='LN'),
+            ffn_num_fcs=2,
+    ):
+
+        super(BaseTransformerLayer, self).__init__()
+        assert set(operation_order) & set(
+            ['self_attn', 'norm', 'ffn', 'cross_attn']) == set(operation_order)
+        num_attn = operation_order.count('self_attn') + operation_order.count(
+            'cross_attn')
+        if isinstance(attn_cfgs, ConfigDict):
+            attn_cfgs = [copy.deepcopy(attn_cfgs) for _ in range(num_attn)]
+        else:
+            assert num_attn == len(attn_cfgs), f'The length ' \
+                f'of attn_cfg {num_attn} is ' \
+                f'not consistent with the number of attention' \
+                f'in operation_order {operation_order}.'
+        assert 'embed_dims' in attn_cfgs[0]
+
+        self.num_attn = num_attn
+        self.feedforward_channels = feedforward_channels
+        self.ffn_dropout = ffn_dropout
+        self.operation_order = operation_order
+        self.act_cfg = act_cfg
+        self.norm_cfg = norm_cfg
+        self.ffn_num_fcs = ffn_num_fcs
+        self.pre_norm = operation_order[0] == 'norm'
+        self.attentions = nn.ModuleList()
+
+        index = 0
+        for operation in operation_order:
+            if operation in ['self_attn', 'cross_attn']:
+                attention = build_attention(attn_cfgs[index])
+                self.attentions.append(attention)
+                index += 1
+
+        self.embed_dims = self.attentions[0].embed_dims
+        self.ffns = nn.ModuleList()
+        num_ffns = operation_order.count('ffn')
+        for _ in range(num_ffns):
+            self.ffns.append(
+                FFN(self.embed_dims, feedforward_channels, ffn_num_fcs,
+                    act_cfg, ffn_dropout))
+
+        self.norms = nn.ModuleList()
+        num_norms = operation_order.count('norm')
+        for _ in range(num_norms):
+            self.norms.append(build_norm_layer(norm_cfg, self.embed_dims)[1])
+
+    def forward(self,
+                query,
+                key,
+                value,
+                query_pos=None,
+                key_pos=None,
+                attn_masks=None,
+                query_key_padding_mask=None,
+                key_padding_mask=None,
+                **kwargs):
+        """Forward function for `TransformerDecoderLayer`.
+
+        Args:
+            query (Tensor): Input query with the shape
+                `(num_query, bs, embed_dims)`.
+            key (Tensor): The key tensor with shape
+                `(num_key, bs, embed_dims)`.
+            value (Tensor): The value tensor with shape
+                `(num_key, bs, embed_dims)`.
+            query_pos (Tensor): The positional encoding for `query`.
+                Default: None.
```

Comment:
Specify the behavior of offering `None` here.

Reply:
In `BaseTransformerLayer`, we can't know the behavior of offering None, because specific process happens in `self_attn` or 
`cross_attn`.